dmutils/status.py

```diff
@@ -47,3 +48,67 @@ def enabled_since(date_string):
         return date_string
 
     return False
+
+
+class StatusError(Exception):
+    """A stub class to use when implementing additional checks for an app's _status endpoint. See API for example."""
+    message = ''
+
+    def __init__(self, message):
+        super().__init__(message)
+        self.message = message
+
+
+def get_app_status(data_api_client=None,
+                   search_api_client=None,
+                   ignore_dependencies=False,
+                   additional_checks=None):
+    """Generates output for `_status` endpoints on apps
+
+    :param current_app: The flask `current_app` global.
+    :param data_api_client: The app's data_api_client, if used.
+    :param search_api_client: The app's search_api_client, if used.
+    :param ignore_dependencies: Minimal endpoint checks only (i.e. the app is routable and disk space is fine).
+    :param additional_checks: A sequence of callables that return dicts of data to be injected into the final JSON
+                              response or raise StatusErrors if they need to log an error that should fail the
+                              check (this will cause the endpoint to return a 500).
+    :return: A dictionary describing the current state of the app with, at least, a 'status' key with a value of 'ok'
+             or 'error'.
+    """
+    error_messages = []
+    core_data = {'status': 'ok'}  # Core checks (endpoint resolves, disk is OK)
+    extra_data = {}  # Data from additional checks to run (if ignore_dependencies is False), e.g. api connectivity
+
+    disk_status = get_disk_space_status()
+    core_data['disk'] = f'{disk_status[0]} ({disk_status[1]}% free)'
+    if disk_status[0].lower() != 'ok':
+        error_messages.append(f'Disk space low: {disk_status[1]}% remaining.')
+
+    if not ignore_dependencies:
+        extra_data['version'] = current_app.config['VERSION']
+        extra_data['flags'] = get_flags(current_app)
+
+        if data_api_client:
+            data_api_status = data_api_client.get_status() or {'status': 'n/a'}
+            extra_data['api_status'] = data_api_status
+            if data_api_status['status'].lower() != 'ok':
+                error_messages.append('Error connecting to the Data API.')
+
+        if search_api_client:
+            search_api_status = search_api_client.get_status() or {'status': 'n/a'}
+            extra_data['search_api_status'] = search_api_status
+            if search_api_status['status'].lower() != 'ok':
+                error_messages.append('Error connecting to the Search API.')
+
+        for additional_check in (additional_checks or []):
+            try:
+                extra_data.update(additional_check())
+
+            except StatusError as e:
+                error_messages.append(e.message)
+
+    if error_messages:
+        core_data['status'] = 'error'
+        core_data['message'] = error_messages
+
+    return jsonify(**{**core_data, **extra_data}), 200 if not error_messages else 500
```

Comment:
If there's a way of doing this that doesn't involve quite so many `**` things then I'd personally find that more readable, but whatevs :)

Reply:
Ha yeah, there is, sure.

Reply:
Done. Actually removed a bit of a weird abstraction by just removing `extra_data` altogether.